source/ways-of-working/standards/db-schema-dev-workflow.html.md.erb

```diff
@@ -0,0 +1,75 @@
+
+![Standard application stack](/images/standard-application-stack.png)
+
+Can the frontend tolerate downtime of backend services? Can the frontend offer reduced functionality when backends are down but still provide a useful service to users? Is the backend 'private' so that it is only accessible to frontend services? If so, then it is possible to perform zero-downtime deploys by taking the backend services down for a short time to perform DB upgrades. This enables 'traditional' database migration deployments. Note, however, that if a backend doesn't support zero-downtime deployments then it must be clearly documented so that any new consuming service knows to be appropriately decoupled.
+
+If the frontend doesn't function without the backend or the 'backend' is also a public-facing service (API) then the backend must also support zero-downtime deployments. If the backend is backed by a defined-schema datastore e.g. an RDBMS then extra steps need to be added to the release pipeline in order to facilitate this.
+
+## 'Traditional' database migration deployments
+- Add new feature to app including a DB schema migration - this is done in-step (single repo containing code and DB update scripts, update contained in single unit of work, app will only work after migration of DB schema)
+
+To deploy:
+
+- stop the application
+- apply DB schema migration scripts (usually managed through a tool such as Flyway or Liquibase)
+- start new version of the application
+
+## Zero-downtime database deployments
+To perform zero-downtime database deployments the schema migration MUST be separated from the code deployment. Database Migrations Done Right explains further and gives an example of the steps required when adding a non-nullable column to an existing table.
+Essentially, database migrations must be done in such a way that they are done independently of application deployments. The underlying principle to obey here is:
+Every change you make must be backward compatible with the rest of the system
+
+For the add non-nullable column scenario these are the backward compatible steps to perform the change:
+
+1. Add nullable column to database – System keeps adding rows, nulls are fine, reads ignore the null
+2. Code change to write correct value to new rows, and handle reading unexpected nulls – Database doesn’t change, now we have some null rows and some rows with data
+3. Run data migration to fill the other columns – This might be a script, or a bit of code in the application, either way your app doesn’t care about any row, it handles data and nulls just fine
+4. Add the non-null constraint – The database now has no nulls and your new code is writing the correct data.
+5. Remove the code that handles the null case – it won’t happen anymore.
+
+![Zero downtime](/images/zero-downtime-db-migrations.png)
+
+## Strategies for refactoring databases
```

Comment:
not sure if referencing a book is the best approach, unlikely someone will go out and read it before doing their db change.

I found a couple of blog references that look quite good and could be added:
- https://medium.com/paypal-tech/postgresql-at-scale-database-schema-changes-without-downtime-20d3749ed680
- https://medium.com/preply-engineering/postgresql-schema-change-gotchas-bf904e2d5bb7

Reply:
Added